Parse gx:Tours from KML

kiselev-dv opened the pull request:
### What's in this PR

* Parse gx:Tour into dataSource.kmlTours array
* Parse LookAt and Camera

### What's not in this PR

* Actual Tours playback*
* Support for `<gx:playMode>` `<gx:TourControl>` `<gx:AnimatedUpdate>`

\* You can find demo implementation of TourPlayer in Sandcastle demo, but properly made implementation should 

* create one combined camera animation tween with whole animation
* play sounds
* determine autoplay from kml document, not by player preferences

hpinkos commented:
Thanks for the update @kiselev-dv!  We'll try to look at this next month after the release comes out

pjcozzi commented:
Thanks again @kiselev-dv!  @tfili do you want to take a first pass over this?

tfili commented on Source/DataSources/KmlLookAt.js:
```diff
@@ -0,0 +1,70 @@
+        Cartesian3,
+        BoundingSphere ) {
+    'use strict';
+
```
Whitespace issues throughout this file. Some functions indent with 2 spaces, some with 4.

tfili commented on Source/DataSources/KmlDataSource.js:
```diff
@@ -1691,6 +1691,141 @@ define([
+    }
+
+    function processTourFlyTo(tour, entryNode) {
+
```
whitespace

tfili commented on Source/DataSources/KmlDataSource.js:
```diff
@@ -1691,6 +1691,141 @@ define([
+        var flyToEntry = { 'entryType': 'FlyTo' };
+
+        var duration = queryNumericValue(entryNode, 'duration', namespaces.gx);
+        if(duration !== null) {
```
Should be `if (defined(duration))`

tfili commented on Source/DataSources/KmlDataSource.js:
```diff
@@ -1691,6 +1691,141 @@ define([
+        }
+
+        var flyToMode = queryStringValue(entryNode, 'flyToMode', namespaces.gx);
+        if(flyToMode !== null) {
```
Use `defined` as well.

tfili commented on Source/DataSources/KmlDataSource.js:
```diff
@@ -1691,6 +1691,141 @@ define([
+        processCamera(entryNode, t);
+
+        if (defined(t.kml.lookAt)) {
+            flyToEntry['lookAt'] = t.kml.lookAt;
```
Should be `flyToEntry.lookAt`.

tfili commented on Source/DataSources/KmlDataSource.js:
```diff
@@ -1691,6 +1691,141 @@ define([
+            flyToEntry['lookAt'] = t.kml.lookAt;
+        }
+        if (defined(t.kml.camera)) {
+            flyToEntry['camera'] = t.kml.camera;
```
Should be `flyToEntry.camera`.

tfili replied:
This is a bunch more places. Fix it throughout.

tfili commented on Source/DataSources/KmlDataSource.js:
```diff
@@ -1691,6 +1691,141 @@ define([
+    }
+
+    function processTourSoundCue(tour, entryNode) {
+        var entry = { 'entryType': 'SoundCue' };
```
Should `entryType` be an enumeration instead?

tfili commented on Source/DataSources/KmlDataSource.js:
```diff
@@ -1691,6 +1691,141 @@ define([
+
+    function processTour(dataSource, parent, node, entityCollection, styleCollection, sourceUri, uriResolver, promises, context) {
+        var name = queryStringValue(node, 'name', namespaces.kml);
+        var tour = {'name': name};
```
We generally write this out like
```
var tour = {
    name: name
};
```

Also note keys don't require quotes.

tfili commented on Source/DataSources/KmlCamera.js:
```diff
@@ -0,0 +1,65 @@
+     * @alias KmlCamera
+     * @constructor
+     */
+    function KmlCamera(feature, options) {
```
Documentation with parameter descriptions.

tfili commented on Source/DataSources/KmlCamera.js:
```diff
@@ -0,0 +1,65 @@
+        this.roll = defaultValue(options.roll, 0);
+    }
+
+    KmlCamera.prototype.getViewOptions = function() {
```
Doc throughout for all new functions.

tfili commented on Source/DataSources/KmlCamera.js:
```diff
@@ -0,0 +1,65 @@
+
+        this._feature = feature;
+
+        this.latitude = defaultValue(options.latitude, 0.0);
```
We should have a member object that is `Cartographic`.

tfili commented on Source/DataSources/KmlCamera.js:
```diff
@@ -0,0 +1,65 @@
+        this.longitude = defaultValue(options.longitude, 0.0);
+        this.altitude = defaultValue(options.altitude, 0.0);
+
+        this.heading = defaultValue(options.heading, 0.0);
```
Use a `HeadingPitchRoll` instance.

tfili commented on Source/DataSources/KmlDataSource.js:
```diff
@@ -1691,6 +1691,141 @@ define([
+        if(defined(camera)) {
+            var result = {};
+
+            result.longitude = queryNumericValue(camera, 'longitude', namespaces.kml);
```
All angles should be stored in radians.

tfili commented on Source/DataSources/KmlDataSource.js:
```diff
@@ -1691,6 +1691,141 @@ define([
+        if(defined(lookAt)) {
+            var result = {};
+
+            result.longitude = queryNumericValue(lookAt, 'longitude', namespaces.kml);
```
Radians.

tfili commented on Source/DataSources/KmlLookAt.js:
```diff
@@ -0,0 +1,70 @@
+
+        this._feature = feature;
+
+        this.latitude = defaultValue(options.latitude, 0.0);
```
Use a `Cartographic` and `HeadingPitchRange` object here as well.

tfili commented on Source/DataSources/KmlLookAt.js:
```diff
@@ -0,0 +1,70 @@
+     * @alias KmlLookAt
+     * @constructor
+     */
+    function KmlLookAt(feature, options) {
```
Add doc throughout the file.

tfili commented:
Thanks @kiselev-dv.

The Sancastle example seems pretty involved. Should some of that functionality be in Cesium itself. I feel like we should have something like a `KmlTour` object that once the datasource is loaded we can find the tour we want and call `Play()` on it. What do you think as far as the general workflow for using a tour @mramato?

kiselev-dv commented:
I will look over notes today or tomorrow. As for `TourPlayer` or ( `KmlTour` ) I was thinking of that it doesn't implement the playback good enough. It just call camera api for actual playback, and I thought maybe It'll worth to try to implement the whole animation as one Camera flight tween, but since KML tours have non blocking features, such as `gx:SoundCue` I'm much more into current approach with just a chain of camera API methods calls.

If you ok with current partial implementation, I'll create `KmlTour` which will encapsulates playback logic.

kiselev-dv commented:
Hi @tfili i've changed the PR, moved TourPlayer functionality to KmlTour, add some tests for it.
If I run tests without `--suppressPassed` and `--failTaskOnError` my tests are marked as passed, but some other test fails
```
1) mergeNewSamples works with huge data sets.
     DataSources/SampledProperty
     Expected spy play to have been called.
    at Specs/DataSources/KmlTourSpec.js:67:28
     Expected spy TourEnd to have been called with [ false ] but it was never called.
    at Specs/DataSources/KmlTourSpec.js:93:29

2) show tile debug labels with memory usage
     Scene/Cesium3DTileset
     Expected 'Texture Memory: 0
Geometry Memory: 0,008' to equal 'Texture Memory: 0
Geometry Memory: 0.008'.
    at Specs/Scene/Cesium3DTilesetSpec.js:1431:62
    at Promise.then (Source/ThirdParty/when.js:196:34)
    at Source/ThirdParty/when.js:297:13
    at processQueue (Source/ThirdParty/when.js:647:4)
    at _resolve (Source/ThirdParty/when.js:333:4)

3) show tile debug labels with all statistics
     Scene/Cesium3DTileset
     Expected 'Geometric error: 70
Commands: 1
Triangles: 120
Features: 10
Texture Memory: 0
Geometry Memory: 0,008
Url: parent.b3dm' to equal 'Geometric error: 70
Commands: 1
Triangles: 120
Features: 10
Texture Memory: 0
Geometry Memory: 0.008
Url: parent.b3dm'.
    at Specs/Scene/Cesium3DTilesetSpec.js:1457:62
    at Promise.then (Source/ThirdParty/when.js:196:34)
    at Source/ThirdParty/when.js:297:13
    at processQueue (Source/ThirdParty/when.js:647:4)
    at _resolve (Source/ThirdParty/when.js:333:4)

4) show only picked tile debug label with all stats
     Scene/Cesium3DTileset
     Expected 'Geometric error: 70
Commands: 1
Triangles: 120
Features: 10
Texture Memory: 0
Geometry Memory: 0,008
Url: parent.b3dm' to equal 'Geometric error: 70
Commands: 1
Triangles: 120
Features: 10
Texture Memory: 0
Geometry Memory: 0.008
Url: parent.b3dm'.
    at Specs/Scene/Cesium3DTilesetSpec.js:1491:58
    at Promise.then (Source/ThirdParty/when.js:196:34)
    at Source/ThirdParty/when.js:297:13
    at processQueue (Source/ThirdParty/when.js:647:4)
    at _resolve (Source/ThirdParty/when.js:333:4)

5) renders multiple instanced tilesets
     Scene/Composite3DTileContent
     Error: Timeout - Async callback was not invoked within timeout specified by jasmine.DEFAULT_TIMEOUT_INTERVAL.
```

If I run them with `--failTaskOnError` I got the error at the end:
```
[19:43:38] Error: 1
    at formatError (/opt/ep/cesium/node_modules/gulp/bin/gulp.js:169:10)
    at Gulp.<anonymous> (/opt/ep/cesium/node_modules/gulp/bin/gulp.js:195:15)
    at emitOne (events.js:96:13)
    at Gulp.emit (events.js:188:7)
    at Gulp.Orchestrator._emitTaskDone (/opt/ep/cesium/node_modules/orchestrator/index.js:264:8)
    at /opt/ep/cesium/node_modules/orchestrator/index.js:275:23
    at finish (/opt/ep/cesium/node_modules/orchestrator/lib/runTask.js:21:8)
    at cb (/opt/ep/cesium/node_modules/orchestrator/lib/runTask.js:29:3)
    at /opt/ep/cesium/gulpfile.js:643:16
    at removeAllListeners (/opt/ep/cesium/node_modules/karma/lib/server.js:379:7)
    at Server.<anonymous> (/opt/ep/cesium/node_modules/karma/lib/server.js:390:9)
    at Server.g (events.js:292:16)
    at emitNone (events.js:91:20)
    at Server.emit (events.js:185:7)
    at emitCloseNT (net.js:1552:8)
    at _combinedTickCallback (internal/process/next_tick.js:77:11)
    at process._tickDomainCallback (internal/process/next_tick.js:128:9)
```
The stacktrace looks quite cryptic, I've got that error now even if I checkout `5588f18` commit which passed some times ago. 

What can I do to debug it further?

kiselev-dv commented:
Hi @hpinkos I think this PR is ready for round two.

hpinkos commented:
@tfili can you take another look?

hpinkos commented:
@kiselev-dv you have some merge conflicts.  Please merge in master when you have a chance

kiselev-dv commented:
@tfili done

kiselev-dv commented:
@hpinkos done

ggetz commented:
Bumped via the forum: https://groups.google.com/forum/#!topic/cesium-dev/Q8cZnejckyo

@kiselev-dv If you bump via GitHub, and especially if you @ mention someone, you'll get a much faster response.

tfili commented on Apps/Sandcastle/gallery/KML Tours.html:
```diff
@@ -0,0 +1,79 @@
+        console.log('Start tour');
+    });
+    tour.tourEnd.addEventListener(function(terminated) {
+        console.log((terminated ? 'Trminate' : 'End') + ' tour');
```
Spelling

tfili commented on Apps/Sandcastle/gallery/KML Tours.html:
```diff
@@ -0,0 +1,79 @@
+        console.log('Play ' + entry.type + ' ('+ entry.duration + ')');
+    });
+    tour.entryEnd.addEventListener(function(entry, terminated) {
+        console.log((terminated ? 'Trminate' : 'End') + ' ' + entry.type);
```
Spelling

tfili commented on Source/DataSources/KmlTourFlyTo.js:
```diff
@@ -0,0 +1,115 @@
+     * @callback KmlTourFlyTo~DoneCallback
+     *
+     * @param {Boolean} terminated true if {@link KmlTourFlyTo#stop} was
+     * caled before entry done playback.
```
Spelling

tfili commented on Source/DataSources/KmlTourWait.js:
```diff
@@ -0,0 +1,54 @@
+     *
+     * @callback KmlTourWait~DoneCallback
+     * @param {Boolean} terminated true if {@link KmlTourWait#stop} was
+     * caled before entry done playback.
```
Spelling

tfili commented:
Sorry about the delay. I sort of lost track of this issue. I'm good with merging this after the few small spelling tweaks.

kiselev-dv commented:
@tfili thanks, done spelling.

@ggetz thanks for advise and pr bump.

cesium-concierge commented:
Congratulations on closing the issue! I found these Cesium forum links in the comments above:

https://groups.google.com/forum/#!topic/cesium-dev/Q8cZnejckyo

If this issue affects any of these threads, please post a comment like the following:
> The issue at https://github.com/AnalyticalGraphicsInc/cesium/pull/5546 has just been closed and may resolve your issue. Look for the change in thenext stable release of Cesium or get it now in the master branch on GitHub https://github.com/AnalyticalGraphicsInc/cesium.

__I am a bot who helps you make Cesium awesome!__ Thanks again.
